README.md
# moogle-shortener

Moogle Shortener is a website that will allow you make URL Shortener.

I make this project as an exercise to improve my clojure skills.
Technologies used:
- Reiit - A fast data-driven router for Clojure(Script).
- Hiccup - A library for representing HTML in Clojure.
- Ring - A Clojure web applications library inspired by Python's WSGI and Ruby's Rack

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.
# moogle-shortener


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-09-06
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-09-06
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/moogle-shortener/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/moogle-shortener/compare/0.1.0...0.1.1


test/moogle_shortener/core_test.clj
(ns moogle-shortener.core-test
  (:require [clojure.test :refer :all]
            [moogle-shortener.core :refer :all]))


project.clj
(defproject moogle-shortener "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.3"]
                 [mount "0.1.16"]
                 [environ "1.2.0"]
                 [org.clojure/tools.logging "1.1.0"]
                 [ch.qos.logback/logback-classic "1.2.5"]
                 [camel-snake-kebab "0.4.2"]
                 [ring/ring-jetty-adapter "1.9.3"]
                 [ring "1.9.4"]
                 [ring/ring-json "0.5.1"]
                 [metosin/reitit "0.5.15"]
                 [commons-validator "1.7"]
                 [hiccup "1.0.5"]
                 [nano-id "1.0.0"]
                 [cheshire "5.10.1"]
                 [borkdude/sci "0.2.6"]
                 [org.xerial/sqlite-jdbc "3.36.0.1"]
                 [com.github.seancorfield/next.jdbc "1.2.689"]]
  :main moogle-shortener.web
  ;:aot :all
  :uberjar-name "moogle-shortener.jar"
  :profiles {:dev     {:source-paths ["dev"]
                       :dependencies [[org.clojure/test.check "1.1.0"]
                                      [ch.qos.logback/logback-classic "1.2.5"]]}}
  :omit-source true
  :resource-paths ["resources"]
  :plugins [[lein-environ "1.2.0"]
            [lein-ring "0.12.5"]]
  :repl-options {:init-ns user}
  :ring {:handler moogle-shortener.web/app})



src/moogle_shortener/db.clj
(ns moogle-shortener.db
  (:require [moogle-shortener.config :as config]
            [next.jdbc.sql :as sql]
            [next.jdbc.result-set :as rs]
            [clojure.tools.logging :as log]))

(def default-row-mapper
  (rs/as-maps-adapter
    rs/as-unqualified-kebab-maps
    rs/clob-column-reader))

(defn insert!
  "Wrapper da funcao next.jdbc/insert! com trace das instruções"
  ([db table key-map]
   (insert! db table key-map {}))
  ([db table key-map opts]
   (log/trace (str "table: [" table "] - key-map: [" key-map "] - opts: [" opts "]"))
   (sql/insert! db
                table
                key-map
                (merge {:builder-fn default-row-mapper} opts))))

(defn query
  "Wrapper of function next.jdbc.sql/query with debug log for all DML."
  ([db sql-params]
   (query db sql-params {}))
  ([db sql-params opts]
   (log/trace (str "sql: [" sql-params "] - params: [" opts "]"))
   (sql/query db
              sql-params
              (merge {:builder-fn default-row-mapper} opts))))


src/moogle_shortener/core.clj
(ns moogle-shortener.core
  (:require [clojure.tools.logging :as log]
            [moogle-shortener.html :as html]
            [nano-id.core :as nano]
            [moogle-shortener.config :as config]
            [moogle-shortener.db :as db]
            [next.jdbc.sql :as sql])
  (:import (java.time Instant)))

(defonce SHORTENED "shortened")

(defn- api-url [{:keys [scheme server-name server-port]}]
  (let [port (when-not (= 80 server-port) (str ":" server-port))]
    (str (name scheme) "://" server-name port)))

(defn main-page [req]
  {:status  200
   :body    (html/greet-page)
   :headers {}})

(defn- save-shortener!
  "Register in the database all relevant info from request and the URL shortened"
  [db register]
  (db/insert! db :shortened register))

(defn save-access!
  "Register in the database all relevant info from URL shortened access"
  [db access]
  (db/insert! db :access access))


(defn shortener [db id]
  (-> (db/query db
                 [(str "select url from " SHORTENED " where id = ?")
                  id])
      first
      :url))

(defn shortener! [db req]
  (log/info "Shortining URL... [" (get-in req [:form-params "url"]) "]")
  (let [base-url (api-url req)
        id (nano/nano-id (:size config/nano))
        url (get-in req [:form-params "url"])
        ip (:remote-addr req)
        register {:id        id
                  :url       url
                  :shortened (str base-url "/" id)
                  :ip        ip
                  :created   (Instant/now)}]
    (save-shortener! db register)
    (html/shortener (assoc register :host base-url))))

(comment
  (nano/nano-id 8)
  prn)

src/moogle_shortener/html.clj
(ns moogle-shortener.html
  (:require [hiccup.page :as page]
            [clojure.java.io :as io]))

(defonce header
         [:head
          [:title "Moogle Url Shortener"]
          [:meta {:name    :viewport
                  :content "width=device-width, initial-scale=1"}]
          [:link {:href "https://cdn.jsdelivr.net/npm/bulma@0.9.3/css/bulma.min.css"
                  :rel  :stylesheet}]
          [:link {:href "https://cdn.jsdelivr.net/npm/bulma-social@2/css/all.min.css"
                  :rel  :stylesheet}]
          [:link {:href "https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css"
                  :rel  :stylesheet}]
          [:link {:href "https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"}]])

(defn greet-page []
  (page/html5 {:lang :en}
              header
              [:body
               [:section {:class "hero-body"}
                [:div {:class "hero-body has-text-centered"}
                 [:h1 {:class "title"} "Moogle Shortener"]]
                [:div {:class "container box"}
                 [:div {:class "field has-addons"}
                  #_[:div {:class "columns is-mobile"}
                     [:div {:class "column is-expanded"}
                      [:figure {:class "image is-32x32 block"}
                       [:img {:src "assets/img/moogle.png"}]]]]]
                 [:p {:class "subtitle"} "Paste the URL to be shortened"]
                 [:form {:method "POST" :action "shortener"}
                  [:div {:class "field has-addons"}
                   [:div {:class "control is-expanded"}
                    [:input {:name "url" :pattern "https?://.+" :class "input is-large" :type "text"}]]
                   [:div {:class "control"}
                    [:input {:class "input is-large" :type "submit"}]]]]]]]))

(defn shortener [register]
  (page/html5 {:lang :en}
              header
              [:body
               [:section {:class "hero-body"}
                [:div {:class "hero-body has-text-centered"}
                 [:h1 {:class "title"} [:a {:href (:host register)} "Moogle Shortener"]]]
                [:div {:class "container box"}
                 [:h1 {:class "title is-4"} "Your shortened URL"]
                 [:h5 {:class "subtitle is-6"} "Copy the shortened link and share it in messages, texts, posts, websites and other locations"]
                 [:div {:class "field has-addons"}
                  [:div {:class "control is-expanded"}
                   [:input {:class "input is-large" :type "text" :value (:shortened register) :read-only true}]]
                  [:div {:class "control"}
                   [:button {:class "button is-success is-large"} "Copy URL"]]]
                 [:p {:class "subtitle is-6"} (str "Long URL: ") [:a {:href (:url register)} (:url register)]]
                 [:br]
                 [:br]
                 [:h2 {:class "title is-h2"} "Share URL"]
                 [:div {:class "buttons"}
                  [:a {:class "button is-medium is-facebook"}
                   [:span {:class "icon"}
                    [:i {:class "fab fa-facebook"}]]
                   [:span "Facebook"]]
                  [:a {:class "button is-medium is-twitter"}
                   [:span {:class "icon"}
                    [:i {:class "fab fa-twitter"}]]
                   [:span "Twitter"]]]
                 [:br]]]]))



src/moogle_shortener/middleware.clj
(ns moogle-shortener.middleware
  (:require [clojure.string :as str]
            [camel-snake-kebab.core :refer [->snake_case_keyword
                                            ->kebab-case-keyword]]
            [camel-snake-kebab.extras :refer [transform-keys]]
            [reitit.ring.middleware.exception :as exception]
            [clojure.tools.logging :as log]
            [moogle-shortener.config :as config])
  (:import (clojure.lang ExceptionInfo)))

(defn wrap-trailing-slash
  "
  This middleware removes the trailing slash in the end of uri.

  In the following example both urls are considered the same:
    https://server:8080/xpto  => {:uri /xpto}
    https://server:8080/xpto/ => {:uri /xpto}
  "
  [handler]
  (fn [req]
    (let [trailing-slash-fn (fn [url]
                              ;; Remove slash in the end of uri
                              ;; Skip when it is the root path
                              (if (> (count url) 1)
                                (clojure.string/replace url #"/$" "")
                                url))
          request (update req :uri trailing-slash-fn)]
      (handler request))))

(defn wrap-db
  "Wraps the database connection"
  [handler]
  (fn [req]
    (handler (assoc req :moogle-shortener/db @config/datasource))))

(defn- bad-request [msg]
  {:status  400
   :headers {}
   :body    {:error   "BAD_REQUEST"
             :message msg}})

(defn- internal-server-error [msg]
  {:status  500
   :headers {}
   :body    {:status 500
             :error   "INTERNAL_SERVER_ERROR"
             :title "Something very bad has happened!"
             :message msg}})

(defn wrap-exception-mapper
  "This middleware adds a default exception handling for all uncaught exceptions.
   Returns a INTERNAL SERVER ERROR response body with the exception message inside.

   WARNING: It must wrap all handlers in order to work correctly."
  [handler]
  (fn [req]
    (try
      (handler req)
      (catch ExceptionInfo ex
        (let [msg (ex-message ex)]
          (log/error ex msg)
          (cond (= (:error (ex-data ex)) :invalid)
                (bad-request msg)

                :else
                (internal-server-error msg))))
      (catch Throwable ex
        (let [msg (ex-message ex)]
          (log/error ex msg)
          (internal-server-error (str ex)))))))


(defn exception-handler
  "Cria mensagem de erro a partir da requisição."
  [type exception request]
  {:status 500
   :body   {:uri       (:uri request)
            :type      type
            :message   (ex-message exception)
            :data      (ex-data exception)
            :exception (class exception)
            :cause     (ex-cause exception)}})

(def exception-middleware
  (exception/create-exception-middleware
    (merge
      exception/default-handlers
      {
       :reitit.coercion/request-coercion  (exception/create-coercion-handler 400)

       :reitit.coercion/response-coercion (exception/create-coercion-handler 500)

       ;; override the default handler
       ::exception/default                (partial exception-handler "default")

       ;; print stack-traces for all exceptions
       ::exception/wrap                   (fn [handler e request]
                                            (log/error e "ERROR" (prn-str (:uri request)))
                                            (handler e request))})))

(defn wrap-kebab-keys [handler]
  (fn [req]
    (let [body-params (transform-keys ->kebab-case-keyword (:body-params req))]
      (log/trace "KEY CONVERTION - REQUEST: " body-params)
      (handler (assoc req :body-params body-params)))))

(defn wrap-snakecase-keys [handler]
  (fn [req]
    (let [response (handler req)
          result (transform-keys ->snake_case_keyword (:body response))]
      (log/trace "KEY CONVERTION - RESPONSE: " result)
      (assoc response :body result))))

(defn wrap-log-request-time [handler]
  (fn [req]
    (log/debug ">>>")
    (log/debug (str/upper-case (name (:request-method req)))
              (:uri req))
    (let [init-time (System/currentTimeMillis)
          response (handler req)
          end-time (System/currentTimeMillis)]
      (log/debug (str/upper-case (name (:request-method req)))
                (:uri req)
                (str "(" (- end-time init-time) "ms)"))
      (log/debug "<<<")
      response)))

(comment
  (transform-keys ->kebab-case-keyword {:abc_dfe {:a-bca "abc"}}))


src/moogle_shortener/web.clj
(ns moogle-shortener.web
  (:require [reitit.ring :as ring]
            [reitit.coercion.malli]
            [reitit.ring.malli]
            [reitit.swagger :as swagger]
            [reitit.swagger-ui :as swagger-ui]
            [reitit.ring.coercion :as coercion]
            [reitit.dev.pretty :as pretty]
            [reitit.ring.middleware.muuntaja :as muuntaja]
            [reitit.ring.middleware.multipart :as multipart]
            [reitit.ring.middleware.parameters :as parameters]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.resource :refer [wrap-resource]]
            [clojure.java.io :as io]
            [muuntaja.core :as m]
            [malli.util :as mu]
            [cheshire.core :as json]
            [moogle-shortener.core :as core]
            [moogle-shortener.middleware :refer [wrap-db
                                                 wrap-trailing-slash
                                                 wrap-kebab-keys
                                                 wrap-snakecase-keys
                                                 exception-middleware
                                                 wrap-log-request-time]]
            [ring.util.response :as response]
            [clojure.tools.logging :as log]
            [moogle-shortener.config :as config]
            [ring.adapter.jetty :as jetty])
  (:import (java.util Properties)
           (java.time Instant)
           (org.apache.commons.validator.routines UrlValidator))
  (:gen-class))

(defn- api-url [{:keys [scheme server-name server-port]}]
  (str (name scheme) "://" server-name ":" server-port "/api"))

(defn- success [body]
  {:status  200
   :headers {}
   :body    body})

(defn- no-content []
  {:status  204
   :headers {}
   :body    nil})

(defn- not-found []
  {:status  404
   :headers {}
   :body    nil})

(defn- accepted [location body]
  {:status  202
   :headers {"Location" location}
   :body    body})

(def handle-show-version
  {:summary ""
   :handler (fn [req]
              (with-open [pom-reader (io/reader (io/resource "META-INF/maven/moogle-shortener/moogle-shortener/pom.properties"))]
                (let [pom (doto (Properties.)
                            (.load pom-reader))]
                  {:status  200
                   :headers {"Content-Type" "application/json"}
                   :body    (json/generate-string pom)})))})

(defn valid-url? [url]
  (let [schemes (into-array String ["http" "https"])]
    (.isValid (UrlValidator. schemes) url)))

(def schema-valid-url?
  (malli.core/-simple-schema
    {:type :user/schema-valid-url?
     :pred #(valid-url? %)}))

(def handle-shortener
  {:summary    ""
   :parameters {:form [:map
                       [:url schema-valid-url?]]}
   :handler    (fn [req]
                 {:status  200
                  :body    (core/shortener! (:moogle-shortener/db req) req)
                  :headers {}})})

(def handle-main-page
  {:summary "Exibe a versão da aplicação em execução"
   :handler core/main-page})

(def handle-redirect
  {:summary "Redirect to the URL shortened"
   :handler (fn [req]
              (let [db (:moogle-shortener/db req)
                    id (get-in req [:path-params :id])
                    access {:id      id
                            :ip      (:remote-addr req)
                            :header  (dissoc (:headers req) "cookie")
                            :created (Instant/now)}]
                (core/save-access! db access)
                (if-let [redirect (core/shortener db id)]
                  (do
                    (log/info "redirecting..." redirect)
                    (response/redirect redirect))
                  (not-found))))})

(comment
  (core/shortener @config/datasource "deH5wiJ6")
  prn)

(def handle-swagger
  {:no-doc  true
   :swagger {:info {:title       "Moogle URL Shortner"
                    :description ""}
             :tags [{:name "api", :description "API Description"}]}
   :handler (swagger/create-swagger-handler)})

(def ^:private router-opts
  {:exception pretty/exception
   :data      {:coercion (reitit.coercion.malli/create
                           {;; set of keys to include in error messages
                            :error-keys       #{#_:type :coercion :in :schema :value :errors :humanized #_:transformed}
                            ;; schema identity function (default: close all map schemas)
                            :compile          mu/closed-schema
                            ;; strip-extra-keys (effects only predefined transformers)
                            :strip-extra-keys true
                            ;; add/set default values
                            :default-values   true
                            ;; malli options
                            :options          nil})
               :muuntaja m/instance}})

(def default-middleware
  [;;custom
   wrap-log-request-time
   ;; query-params & form-params
   parameters/parameters-middleware
   ;wrap-fetch-parameters
   ;; content-negotiation
   muuntaja/format-negotiate-middleware
   ;; encoding response body
   muuntaja/format-response-middleware
   ;; custom
   wrap-snakecase-keys
   ;; custom
   exception-middleware
   ;; decoding request body
   muuntaja/format-request-middleware
   ;; coercing response bodys
   coercion/coerce-response-middleware
   ;; coercing request parameters
   coercion/coerce-request-middleware
   ;; multipart
   multipart/multipart-middleware
   ;; custom
   wrap-kebab-keys
   wrap-db])

(def app
  (ring/ring-handler
    (ring/router
      [["/" {:get        handle-main-page
             :swagger    {:tags ["api"]}
             :middleware default-middleware}]
       ["/shortener" {:post handle-shortener :conflicting true :middleware default-middleware}]
       ["/version" {:get handle-show-version :conflicting true :middleware default-middleware}]
       ["/assets/*" (ring/create-resource-handler)]
       ["/{id}" {:get handle-redirect :conflicting true :middleware default-middleware}]]
      router-opts)
    (ring/routes
      (swagger-ui/create-swagger-ui-handler
        {:path   "/api-docs"
         :config {:validatorUrl     nil
                  :operationsSorter "alpha"}})
      (ring/create-default-handler))))

(defn -main []
  (log/info "Staring App...")
  (jetty/run-jetty app
                   {:port  80
                    :join? false}))

(comment

  (prn))
